kratos/geometries/coupling_geometry.h

```diff
@@ -0,0 +1,373 @@
+        : BaseType( rOther )
+        , mpGeometries( rOther.mpGeometries )
+    {
+    }
+
+    /// Destructor
+    ~CouplingGeometry() override = default;
+
+    ///@}
+    ///@name Operators
+    ///@{
+
+    /**
+     * Assignment operator.
+     *
+     * @note This operator don't copy the points and this
+     * geometry shares points with given source geometry. It's
+     * obvious that any change to this geometry's point affect
+     * source geometry's points too.
+     *
+     * @see Clone
+     * @see ClonePoints
+     */
+    CouplingGeometry& operator=( const CouplingGeometry& rOther )
+    {
+        BaseType::operator=( rOther );
+        mpGeometries = rOther.mpGeometries;
+        return *this;
+    }
+
+    /**
+     * @brief Assignment operator for geometries with different point type.
+     *
+     * @note This operator don't copy the points and this
+     * geometry shares points with given source geometry. It's
+     * obvious that any change to this geometry's point affect
+     * source geometry's points too.
+     *
+     * @see Clone
+     * @see ClonePoints
+     */
+    template<class TOtherPointType>
+    CouplingGeometry& operator=(
+        CouplingGeometry<TOtherPointType> const & rOther )
+    {
+        BaseType::operator=( rOther );
+        mpGeometries = rOther.mpGeometries;
+        return *this;
+    }
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    typename BaseType::Pointer Create(
+        PointsArrayType const& ThisPoints ) const override
+    {
+        return Kratos::make_shared<CouplingGeometry>();
+    }
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+    /**
+    * @brief This function returns the geometry part which is accessable
+    *        with a certain index.
+    * @details Here, 0 is the master, all above 0 is for the slaves
+    * of the geometry. Checks if index is available only in debug mode.
+    * @param Index of the geometry part. 0->Master; all bigger than 1->Slave
+    * @return geometry, which is connected through the Index
+     */
+    GeometryType& GetGeometryPart(IndexType Index) const override
+    {
+        KRATOS_DEBUG_ERROR_IF(mpGeometries.size() <= Index) << "Index "
+            << Index << " out of range. Composite contains only of: "
+            << mpGeometries.size() << " geometries." << std::endl;
+
+        return *mpGeometries[Index];
+    }
+
+    /**
+    * @brief ONLY for coupling_geometry. Not necessary in base class.
+    * @details Allows to exchange certain geometries.
+    * @param Index of the geometry part. 0->Master; 1->Slave
+     */
+    void SetGeometryPart(IndexType Index, GeometryPointer pGeometry)
+    {
+        KRATOS_DEBUG_ERROR_IF(mpGeometries.size() <= Index) << "Index out of range: "
+            << Index << " composite contains only of: "
+            << mpGeometries.size() << " geometries." << std::endl;
+
+        KRATOS_ERROR_IF(0 == Index) << "Cannot change master: "
```

Comment:
Why?

Reply:
The problem is that I would have to update the GeometryData pointer in the base class, which is not possible as it is const.

Reply:
Now I see....